File: app/src/main/java/com/egecius/rxjava2_demo_2/rx/subjects/AsyncSubjectExamples.java
package com.egecius.rxjava2_demo_2.rx.subjects;

import java.util.List;

import io.reactivex.subjects.AsyncSubject;

public class AsyncSubjectExamples {

    static final Exception ASYNC_EXCEPTION = new Exception();

    AsyncSubject<Integer> emitEventsWithOnComplete(List<Integer> list) {

        AsyncSubject<Integer> subject = AsyncSubject.create();

        for (final Integer integer : list) {
            subject.onNext(integer);
        }

        subject.onComplete();

        return subject;
    }

    AsyncSubject<Integer> emitEventsWithError(List<Integer> list) {

        AsyncSubject<Integer> subject = AsyncSubject.create();

        for (final Integer integer : list) {
            subject.onNext(integer);
        }

        subject.onError(ASYNC_EXCEPTION);

        return subject;
    }

    AsyncSubject<Integer> emitEventsWithoutTermination(List<Integer> list) {
        AsyncSubject<Integer> subject = AsyncSubject.create();

        for (final Integer integer : list) {
            subject.onNext(integer);
        }

        // no terminate event

        return subject;
    }
}

File: app/src/main/java/com/egecius/rxjava2_demo_2/rx/subjects/AsyncSubjectExamples.kt
package com.egecius.rxjava2_demo_2.rx.subjects

import io.reactivex.subjects.AsyncSubject

class AsyncSubjectExamples {

    fun emitEventsWithOnComplete(list: List<Int>): AsyncSubject<Int> {

        val subject = AsyncSubject.create<Int>()

        for (integer in list) {
            subject.onNext(integer)
        }

        subject.onComplete()

        return subject
    }

     fun emitEventsWithError(list: List<Int>): AsyncSubject<Int> {

        val subject = AsyncSubject.create<Int>()

        for (integer in list) {
            subject.onNext(integer)
        }

        subject.onError(ASYNC_EXCEPTION)

        return subject
    }

     fun emitEventsWithoutTermination(list: List<Int>): AsyncSubject<Int> {
        val subject = AsyncSubject.create<Int>()

        for (integer in list) {
            subject.onNext(integer)
        }

        // no terminate event

        return subject
    }

    companion object {

        val ASYNC_EXCEPTION = Exception()
    }
}
